drivers/bestfortress.c

```diff
@@ -420,42 +499,60 @@ static int upsdrv_setvar (const char *var, const char * data) {
 	return STAT_SET_HANDLED;
 }
 
+/*
+ * The "power down and maybe return command" is "OFF %d\r", with a
+ * delay in seconds before poweroff.  As a special case, "OFF 0" does
+ * not shut down.  The UPS will power on the load when power returns
+ * (or after a delay if power is not out), according to the front panel
+ * parameter, or the value set via `autorestart()`.
+ */
+
+/*
+ * This is equivalent to the `shutdown.return` instant command, but
+ * invoked with `-k`.
+ * \todo Reduce duplication.
+ */
 void upsdrv_shutdown(void)
 {
 	const	char	*grace;
 
+	upsdebugx(2, "%s: begin", __func__);
+
 	grace = dstate_getinfo("ups.delay.shutdown");
+	if (!grace) {
+		upsdebugx(1, "%s: ups.delay.shutdown is NULL!", __func__);
+		/* Pick a different value than 20 so we can see it in the logs. */
+		grace = "30";
```

Comment:
Not sure about losing `grace = "1"; /* apparently, OFF0 does not work */` here for eternity.

Probably the original intent was "if not set, go down immediately" which then got confused by a default setting (20) which comes later in process lifecycle when not `-k`. Maybe it is that default that should go away, eventually, and users would be expected to give a setting in `ups.conf`?..

Reply:
I moved the comment about OFF0 not working to higher up where I explain the protocol.   That's where it belongs, rather than as a side comment about something else.

Now, the code uses the configured delay because I fixed the ordering bug you found for me.  The 30 is just  a "if the impossible try to cope" and I picked 30 vs 20 so that it might be noticeable.

IMHO, it *never* made sense to use 1 when the variable was empty.  That's essentially always too hasty.

Reply:
Ok, makes sense. Just making sure it is a conscious decision :)